third_party/amd/lib/TritonAMDGPUTransforms/OptimizeBufferOpPtr.cpp

```diff
@@ -0,0 +1,323 @@
+      auto denseAttr =
+          dyn_cast<DenseIntElementsAttr>(constantOp.getValueAttr());
+      if (!denseAttr)
+        return false;
+      if (!denseAttr.isSplat())
+        return false;
+      return true;
+    } else if (auto splat = dyn_cast<triton::SplatOp>(defOp)) {
+      return isScalarizableValue(splat.getSrc());
+    }
+    return false;
+  }
+
+  // Generate a scalar value that is stored in provided RankedTensor val
+  static mlir::Value scalarizeValue(PatternRewriter &rewriter,
+                                    mlir::Value val) {
+    if (isa<mlir::IntegerType>(val.getType()))
+      return val;
+    auto defOp = val.getDefiningOp();
+    assert(defOp);
+    if (auto constantOp = dyn_cast<mlir::arith::ConstantOp>(defOp)) {
+      auto denseAttr =
+          dyn_cast<DenseIntElementsAttr>(constantOp.getValueAttr());
+      assert(denseAttr);
+      auto splatVal = denseAttr.getSplatValue<llvm::APInt>();
+      rewriter.setInsertionPoint(constantOp);
+      return rewriter.create<mlir::arith::ConstantIntOp>(
+          constantOp.getLoc(), denseAttr.getElementType(), splatVal);
+    } else if (auto splat = dyn_cast<triton::SplatOp>(defOp)) {
+      return scalarizeValue(rewriter, splat.getSrc());
+    }
+    return Value();
+  }
+
+  // Description of struct fields:
+  // - op is a target BufferOp, for example BufferLoadOp or BufferLoadToLocalOp
+  // - offsetIncrement is a tensor added to offsets on each iteration
+  // - baseIncrement is a scalar which will be added to base pointer
+  // - offsetInitializer is a value of offset on first loop iteration
+  // - incrementOp is an operation that advances offset tensor
+  struct BufferOpInfo {
+    amdttg::BufferOpInterface op;
+    Value offsetIncrement;
+    Value baseIncrement;
+    Value offsetInitializer;
+    Operation *incrementOp;
+  };
+
+  // Perform series of checks to decide if given operation could be optimized.
+  // If optimization is possible, return filled BufferOpInfo
+  static std::optional<BufferOpInfo>
+  analyzeBufferOp(amdttg::BufferOpInterface op, scf::ForOp targetFor) {
+    LDBG("Analyzing: " << *op);
+    Value maybeOffsetsBlockArg = op.getOffsets();
+    auto maybeOffsetDefOp = maybeOffsetsBlockArg.getDefiningOp();
+    if (maybeOffsetDefOp && isa<arith::AddIOp>(maybeOffsetDefOp)) {
+      for (auto &use : maybeOffsetDefOp->getUses()) {
+        auto yieldOp = dyn_cast<scf::YieldOp>(use.getOwner());
+        if (!yieldOp || yieldOp->getParentOp() != targetFor) {
+          continue;
+        }
+        auto loopBody = targetFor.getBody();
+
+        int blockOpNo =
+            use.getOperandNumber() + targetFor.getNumInductionVars();
+        maybeOffsetsBlockArg = loopBody->getArgument(blockOpNo);
+        break;
+      }
+    }
+    if (!isa<BlockArgument>(maybeOffsetsBlockArg)) {
+      LDBG("Rejected: expect buffer op offset to be a loop argument");
+      return {};
+    }
+    auto blockArg = dyn_cast<BlockArgument>(maybeOffsetsBlockArg);
```

Comment:
You can move this to be ahead of L139 so in L139 you can check `if (!blockArg)`.